### ISSUE TITLE
netrw's "gx" functionality doesn't open URLs anymore in macOS

### ISSUE THREAD

**Issue by ychin** [vim/vim#4738]
**Describe the bug**
The latest Vim has updated to a newer version of netrw, which seems to have broken the `gx` functionality. In the past, positioning the cursor on an URL and press `gx` would open the URL in a web browser (using the `open` command). The updated netrw would instead try to curl it, which seems wrong.

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Run `vim --clean` (or `gvim --clean`, etc.) in macOS.
2. Enter insert mode and type "https://www.vim.org"
3. Go to normal mode and type `gx`.
4. Observe that Vim doesn't open the URL in a web browser.

**Expected behavior**
Vim should open the URL in a web browser.

**Screenshots**
If applicable, copy/paste the text or add screenshots to help explain your problem.

**Environment (please complete the following information):**
 - Vim version [e.g. 8.1.1755]
 - OS: macOS 10.14 (haven't tested others yet)

**Additional context**
Bug was introduced in 85850f3a5ef9f5a9d22e908ef263de8faa265a95. Originally I got this bug filed against MacVim at https://github.com/macvim-dev/macvim/issues/925 so cross-referencing here.

This is a netrw bug so I'm not sure if this is the correct place to post it, but with the ownership question I'm not 100% sure who owns netrw now and where I should file a bug for it to be fixed and updated to Vim's bundled plugin.

**Comment by brammool** [comment]
Can you make it work by setting  g:netrw_http_cmd ?

**Comment by eirnym** [comment]
For me this command opens some random file from the history, despite setting of `g:netrw_http_cmd`. The conditions are the same.

**Comment by ychin** [comment]
I did `let g:netrw_http_cmd='open'` and it didn't seem to work for me. I could take a closer look too if no one takes it up. But I think the ideal case is that the old behavior is preserved, as `open` is the default way to open generic paths and has been working before.

**Comment by UncleBill** [comment]
I got same issue.

**Comment by cecamp** [comment]
I don't have a mac, and I believe that I'm no longer the maintainer.  However, try using g:netrw_browsex_viewer, not g:netrw_http_cmd.  If that doesn't help, then refer to the directions in   :help netrw_debug  .  Perhaps the macunix exception should be pushed before the has("unix") tests in netrw#BrowseX().

**Comment by mipmip** [comment]
I got same issue.

**Comment by mcepl** [comment]
I got the same problem with neovim (building from master on Linux/openSUSE). With ``g:netrw_browsex_viewer`` set to ``setsid osurl`` (my own script preprocessing references, but the same thing happens when using normal ``xdg-open``, which is the equivalent of Mac’s ``open``) vim downloads the page from the Web with curl and then opens in the browser the downloaded file (which is completely broken of course).

Any idea how to debug it?

---

Damn! And now, I couldn’t reproduce it again, without changing anything anywhere. Sorry, no idea, what happened.

**Comment by mcepl** [comment]
I have my deep suspicion about https://github.com/vim/vim/blob/master/runtime/autoload/netrw.vim#L5309. That is incorrect, isn’t it? At least, when I run ``:echo g:netrw_browsex_viewer =~ '\s'`` I get 1.

**Comment by mcepl** [comment]
Of course, it is wrong. When my ``g:netrw_browsex_viewer`` equals ``'setsid osurl'`` (which exactly corresponds to the examples in the documentation), netrw thinks I have that variable empty.

**Comment by cecamp** [comment]
Anyone willing to use the directions in  :help netrw-debug  with the gx on a mac problem?  I myself don't have a mac and so I can't debug this issue without some help.

**Comment by ngortheone** [comment]
this happens on linux too

**Comment by brammool** [comment]
> this happens on linux too

What I notice:
- File written to a temp file (how is that ever cleaned up?)
- File name is a number, browser only recognizes it as HTML when adding
  the .html extension.
- It closes the current buffer and goes back to the previous one


-- 
'Well, here's something to occupy you and keep your mind off things.'
'It won't work, I have an exceptionally large mind.'
		-- Douglas Adams, "The Hitchhiker's Guide to the Galaxy"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mcepl** [comment]
I see! So, only truly ``http[s]?:`` URLs are broken and result in curl craziness. When running ``gx`` on something weird (e.g., ``gh#vim/vim#4738``, which is OK, my script will translate it into URL of this ticket), then [my script](https://gitlab.com/mcepl/vim-suse-changes/blob/master/osurl) is run and everything works just fine.

So, apparently, netrw somehow got mixed up *editing* of the file from given URL, and *jumping* to it with ``gx``. ``:e https://matej.ceplovi.cz/blog/harry-potter-and-aristotle.html`` works as expected.

**Comment by ngortheone** [comment]
I am on nvim
@mcepl 

```
:e http://info.cern.ch
```
opens a new split and gives me

```
**error** (netrw) not a netrw-style url; netrw uses protocol://[user@]hostname[:port]/[path])
```

`gx` on the same link opens a new split, where this message flashes and disappears, leaving a split empty 



And your example `:e https://matej.ceplovi.cz/blog/harry-potter-and-aristotle.html` opens html of the page

`gx` on the same link downloads the html file, and the opens it in a browser, so I see this URL
`file:///tmp/nvimOl8arl/harry-potter-and-aristotle.html` in path string

**Comment by ngortheone** [comment]
So I think there is a bug somewhere in URL parsing that leads to downloading pages instead of simply opening them

**Comment by mcepl** [comment]
> And your example `:e https://matej.ceplovi.cz/blog/harry-potter-and-aristotle.html` opens html of the page
> 
> `gx` on the same link downloads the html file, and the opens it in a browser, so I see this URL
> `file:///tmp/nvimOl8arl/harry-potter-and-aristotle.html` in path string

That’s exactly what I see. ``:e URL`` opens HTML of the page (which is what’s expected, isn’t it?) and ``gx`` screws things up.

**Comment by jottkaerr** [comment]
> I am on nvim
> @mcepl
> 
> ```
> :e http://info.cern.ch
> ```
> 
> opens a new split and gives me
> 
> ```
> **error** (netrw) not a netrw-style url; netrw uses protocol://[user@]hostname[:port]/[path])

It seems `:e URL` needs at least a slash in the url to separate the host name from the possibly empty path. So
```
:e http://info.cern.ch/
```
should work and open the html in Vim (at least it does for me).

**Comment by mcepl** [comment]
> It seems `:e URL` needs at least a slash in the url to separate the host name from the possibly empty path.

There are mountains of literature about that slash, and how ``URL/`` is quite certainly not the same thing as ``URL``. So, if netrw doesn’t consider them as the same thing, it is actually The Right Thing™.

**Comment by 9mm** [comment]
+1 subscribing to know when this is figured out, dis curl is cramping my style

**Comment by chrisbra** [comment]
If you want to help fix this, how about going with the suggestions by [Charles](https://github.com/vim/vim/issues/4738#issuecomment-519215989) (e.g. provide the information from `:h netrw-debug`?)

**Comment by mcepl** [comment]
@cecamp Here it is.
[DBG.txt](https://github.com/vim/vim/files/3490385/DBG.txt)

I have just run
```sh
 $ vim -u /tmp/netrw.vimrc ~/build/home\:mcepl\:branches\:devel\:languages\:python\:Factory/python-rpm-macros/python-rpm-macros.spec
```
got a cursor on the line with URL and pressed ``gx``.

**Comment by chrisbra** [comment]
cc @cecamp

**Comment by chrisbra** [comment]
Alternative mentioned by Ken in #1386 (https://github.com/vim/vim/issues/1386#issuecomment-351320760) is to use [open-browser](https://github.com/tyru/open-browser.vim)

**Comment by eirnym** [comment]
It's not a bad idea to replace usage of bundled plugin with third-party, but it doesn't solve the problem

**Comment by eirnym** [comment]
Does this plugin have high level tests (aka integration tests) to coverup underlying changes and tell that it works the same as it did before? 

I ask as I didn't found any

**Comment by victoriastuart** [comment]
Workaround, based on comments at https://stackoverflow.com/a/20177492/1904943 and other sources.  Add to this to your `~/.vimrc`:

```
nmap gx yiW:!xdg-open <cWORD><CR> <C-r>" & <CR><CR>
```
`<cWORD>` should take care of URLs with hashtags (`#`).

Here is my understanding of this mapping.

Per discussion at https://stackoverflow.com/questions/10862457/what-does-c-r-means-in-vim/10863134 `<C-r>` is like doing `CTRL+R` on the keyboard. `<CR>` of course is like hitting `<Enter>`.

`yiW` (note capital `W`) yanks the URL to the unnamed/default register (`"`).

See also https://vi.stackexchange.com/questions/2801/how-can-i-make-gx-recognise-full-urls-in-vim re: `<cWORD>`, etc.

---

Edit: I am on Linux; I just noted the Issue title is ... macOS, so perhaps use `open` in place of `xdg-open` if you are on that OS?  https://superuser.com/questions/911735/how-do-i-use-xdg-open-from-xdg-utils-on-mac-osx

**Comment by blueyed** [comment]
This is caused by 85850f3a5.

A difference from there is that `netrw#BrowseX` is called with `a:remote = 1` from the mapping:
```vim
  nno <silent> <Plug>NetrwBrowseX :call netrw#BrowseX(netrw#GX(),netrw#CheckIfRemote(netrw#GX()))<cr>
```

Old:
```vim
  nno <silent> <Plug>NetrwBrowseX :call netrw#BrowseX(expand((exists("g:netrw_gx")? g:netrw_gx : '<cfile>')),netrw#CheckIfRemote())<cr>
```

**Comment by cecamp** [comment]
Please try v166a of netrw.vim.  If that doesn't work for mac users, please look at the help for g:netrw_http_cmd and find a command that works; then let me know and I'll put it in as default for macos.  You can get v166a from my website, http://www.drchip.org/astronaut/vim/index.html .

**Comment by blueyed** [comment]
@cecamp 
I've tried v166a (which would be easier if it would be available via Git directly btw), but it still does not work (on Linux).
The (main) problem is not mac, but that it considers the buffer to be a remote file already.
This patch works for me: https://github.com/vim/vim/pull/4894/files (and was merged for Neovim already, https://github.com/neovim/neovim/pull/10938).

**Comment by chrisbra** [comment]
the change from #4894 has now been included as of https://github.com/vim/vim/commit/589edb340454e7f1b19358f129287a636d53d0e1#diff-532bf738932796a9a60951fb92334bac

**Comment by cecamp** [comment]
The "solution" given essentially is removing "if filetype is netrw" handling - not good; and making "netrw#CheckIfRemote" return a falsehood (ie. that http:... is not a remote file), also not good.  For this lying effect only the second call to netrw#GX() in the macro needed to be removed, by the way, thus preserving the "if filetype is netrw" handling.  It was intended that netrw should return something editable; ie. the html for the page.  I'm getting the html for the page (under linux), which was the desired behavior and is consistent with what happens with ":e http://..." handling.  Setting g:netrw_browsex_viewer to "firefox" would bring up the page in a browser.  I could retain the correct behavior (ie. returning html) but allow an option to use a browser for http[s]:// urls (where the user would be responsible for picking "firefox", "seamonkey", "iexplorer", whatever).  How about g:netrw_browser or g:netrw_browsex_browser?

**Comment by mcepl** [comment]
@cecamp Couple of notes:

1. First of all, I have never claimed this to be perfect solution, or that I would actually understand what’s going on in netrw. It just removed a regression for me, that’s all.

2. Why does it matter that much, whether I press ``gx`` over correct URL or something else? I have ``g:netrw_browsex_viewer`` set to ``'setsid osurl'``, where ``osurl`` is [this Lua script](https://gitlab.com/mcepl/vim-suse-changes/blob/master/osurl). It basically means, that when I press ``gx`` over some [SUSE abbreviated URLs](https://en.opensuse.org/openSUSE:Packaging_Patches_guidelines#Current_set_of_abbreviations) I get my default browser called with proper URL. So, for example ``gh#vim/vim#4738`` gets me to this issue, ``boo#1151490`` gets me to [this bug in openSUSE bugzilla](https://bugzilla.opensuse.org/show_bug.cgi?id=1151490). This is perhaps in the realm of [XKCD 1172](https://xkcd.com/1172/), but it would be awesome if this post-processing URL functionality was preserved.

**Comment by SRGOM** [comment]
For those suggesting xdg-open, it's itself an unusable- you can never get it to use the program you actually want. I've tried manually editing files in multiple different locations, linking them, using that fabled but useless perl script and what not. The best part is that of the 3 files it looks for in a location, one of the files is deleted and recreated upon change. I've never seen a worse program in the computer world than the old xdg-defaults mechanism

**Comment by eirnym** [comment]
I bet there's a misunderstanding in definition of this plugin, it's usage and the further expansion of it's functionality.

There's nothing wrong that people use it to just open urls. In this case viewer shouldn't download them, like @mcepl suggests.
There's nothing wrong that people wants to edit remote files (download, edit and then upload). In this case downloading and uploading process should be separate from viewing, though. This is like @cecamp suggests.

Could this plugin do both? definitely yes, but not with the same `gx` command

**Comment by mcepl** [comment]
> For those suggesting xdg-open, it's itself an unusable- you can never get it to use the program you actually want.

Are you on Linux? Which distribution you use? What is the bug number of the issue you raised with the maintainers of xdg-utils in your distribution? I mean, fixing bugs is a way more useful (and enjoyable experience) than ranting about unusable programs.

**Comment by chrisbra** [comment]
I had problems with xdg-open in the past. I have been trying to debug this and it was indeed complicated shell scripts, hard to find out what it does. Plus, it does not look very-well maintained, at least I never got any reactions on the bug report I posted: https://bugzilla.xfce.org/show_bug.cgi?id=12251 No reactions for 4 years and "Resolved Invalid" :(

**Comment by mcepl** [comment]
> Plus, it does not look very-well maintained, at least I never got any reactions on the bug report I posted: https://bugzilla.xfce.org/show_bug.cgi?id=12251 No reactions for 4 years and "Resolved Invalid" :(

https://freedesktop.org/wiki/Software/xdg-utils/ is rather enlightening reading. You would find out that the proper place to file the ticket is currently https://gitlab.freedesktop.org/xdg/xdg-utils/issues (it used to be somewhere in https://bugs.freedesktop.org; or perhaps your distro’s bugzilla would work), certainly not the XFCE bugzilla.

**Comment by chrisbra** [comment]
The [actual symptom was that gx did not work from within Vim](https://vi.stackexchange.com/questions/5032/gx-not-opening-url-in-gvim-but-works-in-terminal/5034#5034). Before opening a bug report for netrw, I decided to do my best and try to find out what was wrong and how it could be fixed.

So I did trace netrw to find out, that netrw was basically calling `xdg-open`. So I checked `:!xdg-open` which had the exact same problem. So before opening a bug report for xdg-open I decided to trace xdg-open to find out what was happening, just to find out, that xdg-open does call `exo-open` in a xfce session. 

So I tested using `:!exo-open` which did have the exact same problem. So I checked who is the upstream repository to be able to open a bug report at the correct repository in order to have this fixed.

That's why I opened a bug report at the XFCE desktop bugzilla and not for the freedesktop issue reporting issue. And if you would have read the bugzilla entry, you would have seen that the issue was clearly about `exo-open`.

So I did my duties to be helpful and report it at the proper place and provide an easy to reproduce example. And that place is neither the netrw bug tracker (e.g. here) or the freedesktop repository you mentioned, but the XFCE bugzilla, don't you think?

**Comment by mcepl** [comment]
> So I did my duties to be helpful and report it at the proper place and provide an easy to reproduce example. And that place is neither the netrw bug tracker (e.g. here) or the freedesktop repository you mentioned, but the XFCE bugzilla, don't you think?

You did, I haven’t spent enough time trying to understand that bug, and thought the bug went other way around (exo-open was broken, because xdg-open was broken). Then however, why do you complain about xdg-open, if that was innocent in the problem?

Besides, ``xdg-open`` as used in my script is completely not the point, and I think it is as good default as anything else (do you have anything more default)? If it doesn’t work for you, set ``g:netrw_browsex_viewer`` to something else.

**Comment by 9mm** [comment]
![](https://media0.giphy.com/media/tyqcJoNjNv0Fq/giphy.gif?cid=790b76114d1a65ed2284eeb6a8beff8aa76144cc0c0ba0dd&rid=giphy.gif)

**Comment by SRGOM** [comment]
@mcepl Chris is being patient and polite but I've wasted too hours in my life with semi-maintained programs that I know fully well what works and what to do when it doesn't work . Xdg defaults mechanism is the crappiest piece of software I've seen across all 3 major desktop operating systems. 

I challenge ou to write a succinct way to do one thing-

I want a single file (or Max two if this would involve setting mimetypes)  that can be version controlled and `stow`ed in my home-dir which will set default programs for extensions. I want it to take priority over anything else. I don't care to query/change using commands  . Come back when you have this. 

Edited

**Comment by jamessan** [comment]
I use defaults.list.  As an example:

```
$ cat ~/.local/share/applications/defaults.list
application/pdf=zathura.desktop
text/html=firefox.desktop
x-scheme-handler/http=firefox.desktop
x-scheme-handler/https=firefox.desktop
```

See also [xdg-mime](https://portland.freedesktop.org/doc/xdg-mime.html), which seems to be complementary (it recognizes the defaults.list data, but `xdg-mime default ...` modifies a different file).

**Comment by mgedmin** [comment]
\~/.local/share/applications/defaults.list is for applications you install into \~/.local (e.g. ~via flatpak user installs~ nah flatpak has its own entry in $XDG_DATA_DIRS, but apps manually installed via ./configure --prefix=~/.local/ && make install) to tell about what files they can handle.

\~/.config/mimeapps.list is for user preferences meant to override defaults.list.

**Comment by SRGOM** [comment]
So can I version check-in defaults.list and then link ~/.local/share/applications/defaults.list to that? I remember that there was at least one file out of the 799 that xdg-defaults reads which was deleted and rewritten. I hope defaults.list isn't that? I can mark the other 798 read only if only it helps me get consistent behaviour . Or just edit whichever is the highest priority 

EDIT: 
Against what I noted long time back in my env file, I decided to waste some time again today. 
```
common/functions
191:  echo "xdg-open is cancer. Don't waste type"
```

First thing I noted is that the `xdg-open x.pdf` opened pdf in inkscape. Good. Let's find references to inkscape. `cd  ~/.config; rg -i inkscape `. Nothing. `cd ~/.local; rg -i inkscape`. Nothing again. Nopes, not going to waste further time. Fine I wasted some. Added `application/pdf=zathura.desktop` to `~/.config/mimeapps.list `, `xdg-open` still opens with inkscape.

@mgedmin 's answer , while seemingly confident is (obviously, because it's xdg-open you're dealing with) wrong. See https://unix.stackexchange.com/questions/267020/front-end-to-config-mimeapps-list . Also since he downvoted, I'm going to balance it out with a high-school like self-upvote.

**Comment by Shane-XB-Qian** [comment]
sorry jump in here for an old ticket, though the issue looks still there.
to who do not like the default behavior which download content first but not open url directly.

you can :
1, let g:netrw_nogx = 1
2, re-map gx to some cmd you like, or simply wrote a vimscript or python script or whatever you like.
3, done.

:) btw: i do not like that default behavior either.

**Comment by k-takata** [comment]
> 2, re-map gx to some cmd you like, or simply wrote a vimscript or python script or whatever you like.

As I wrote in another issue, I recommend [open-browser](https://github.com/tyru/open-browser.vim) for this purpose.

**Comment by Shane-XB-Qian** [comment]
cool, so the issue probably can be closed. :+1:

**Comment by chrisbra** [comment]
no, because it is still an issue with the default netrw plugin.

**Comment by ychin** [comment]
Hi, sorry to stir an old bug up, but seems like `gx` for URL's is broken again after November (5ef1c6a483), when Netrw was updated to a new version (v166) and therefore removing the patch. Now it just seems to do something odd (I'm on macOS) and download a temp file and try to open something in the browser but not the website I want.

Is it possible to re-apply the patch as mentioned in https://github.com/vim/vim/issues/4738#issuecomment-533543626?

I guess the other question I have is I read through the thread and it seems like there's some philosophical design reason and benefit that necessitated breaking `gx` (which is a regression because it used to work which I feel is antithetical to the Vim philosophy of not breaking existing keybindings/features) but I'm not sure what it is.

**Comment by boykoatwork** [comment]
Write small plugin for replace gx functionality in Vim. It's still raw and dirty, but...

https://github.com/tengucrow/interceptor

**Comment by tengucrow** [comment]
Write small plugin for replace gx functionality in Vim. It's still raw and dirty, but...

https://github.com/tengucrow/interceptor

**Comment by ghost** [comment]
Hi,
I have this issue on Archlinux. I compiled from the sources then, the issue is also in HEAD.
For me, latest upstream release **v8.0.1850** is working
The very next one: **v8.1.0000**  has issue with gx.
but with another message this time:
`this version of netrw requires vim v7.4 with patch 213`

But I don't see code changes for netrw between those two version, no obvious one. (i am not a reference, pls check)

I don't have this issue with neovim, though.
My terminal interprets the urls so, I could just click but the idea is to fix a regression, isn't it?

thank you!

**Comment by k-takata** [comment]
> The very next one: **v8.1.0000** has issue with gx.
> but with another message this time:
> `this version of netrw requires vim v7.4 with patch 213`

It is a totally different issue and already fixed in v8.1.0001.

**Comment by ghost** [comment]
> > The very next one: **v8.1.0000** has issue with gx.
> > but with another message this time:
> > `this version of netrw requires vim v7.4 with patch 213`
> 
> It is a totally different issue and already fixed in v8.1.0001.

I might express myself incorrectly.  The issue is still gx unable to open link in external browser and messing around with buffers.

v8.0.1850 is working
everything after this version were failing. v8.1.0000 is the first one with this message but it does not matter. Since v8.1.0000 , gx is not working.

I have just tested with the new HEAD, same issue: gx seems to download the content into a temporary file and open it with your $WEBROWSER which is messed up because of no css. 
Then going back to vim with another buffer: sometime empty, sometime the content of the html page, sometime an old buffer...

**Comment by chrisbra** [comment]
please check your version of netrw (e.g. `:e $VIMRUNTIME/autoload/netrw.vim`). And please also check with the latest netrw version from Charles: http://www.drchip.org/astronaut/vim/index.html#NETRW

Thanks.

**Comment by ghost** [comment]
Hello Chris,

It just crossed my mind actually  and yes, the version included in vim's GIT repo is v168. I have used the one from vim v8.0.1850 ( netrw.vim v156 )  and  neovim (v165) and it works. I see that Charles's version is even more recent, v170 (7th June 2020), did not test it.

Good version so far: v156, v165 
Why the embedded plugins are not in sync with upstream ? Actually, why such plugins are not developed in the same codebase ?
It is pretty admitted that netrw is part of VIM.

Of course, I could replace the gx with another custom call but the regression here is quite problematic. 

Thank you again, chris!

**Comment by cecamp** [comment]
I don't have access to a mac anymore, and netrw v170g works just fine with ftp under linux.  Please look at :help netrw-debug and get a debugging trace of what happens when you use gx; you then may send it to me if you wish.

**Comment by Shane-XB-Qian** [comment]
> vim v8.0.1850 ( netrw.vim v156 ) and neovim (v165)

just a fyi: though it was same version num as v165 in neovim, but probably that was a modified version of netrw in it, we'v seen neovim modified netrw, or unless it had rollbacked later then.
// i do not have neovim, but just discussed it at some ticket some months ago.

**Comment by cheap-glitch** [comment]
Can we get a fix for this? It seems updating netrw to the latest version (v170) in the repo would solve this.
Thanks!

**Comment by slonik-az** [comment]
After more than one year of being open and generating significant user attention the bug is still present in vim-8.2.1456 (macvim-165). It is so annoying!

**Comment by k-takata** [comment]
Netrw v170 is now included. https://github.com/vim/vim/commit/1d59aa1fdfb191d9872ff87eb94652acd374b293#diff-3abb9629535900f0c07a312231e18483R4
Closing.

**Comment by slonik-az** [comment]
@k-takata : The "gx" bug seems to be still present. Please, see https://github.com/macvim-dev/macvim/issues/1094 for details.

**Comment by k-takata** [comment]
Hmm, reopening then.

**Comment by strboul** [comment]
A small workaround until it's fixed ([adapted from](https://stackoverflow.com/questions/9458294/open-url-under-cursor-in-vim-with-browser/)):

```vim
if has('macunix')
  function! OpenURLUnderCursor()
    let s:uri = matchstr(getline('.'), '[a-z]*:\/\/[^ >,;()]*')
    let s:uri = shellescape(s:uri, 1)
    if s:uri != ''
      silent exec "!open '".s:uri."'"
      :redraw!
    endif
  endfunction
  nnoremap gx :call OpenURLUnderCursor()<CR>
endif
```

**Comment by 9mm** [comment]
@strboul dude, amazing. This has been literally driving me insane for months.

**Comment by k-takata** [comment]
Okay, I also confirmed that this has been broken also on Windows since Netrw v165.
The problem is that Netrw tries to download the specified URL first, then execute the platform specific open command ("open" on macOS, "rundll32" on Windows) even the command can handle a URL directly.

Before v165, it seems that a URL was somehow determined as not a remote file, then `netrw#BrowseX()` worked nice unexpectedly.

**Comment by k-takata** [comment]
I created PR #7188.
@cecamp Please check this.

**Comment by grr** [comment]
looks like this is stalled. how do we get this pushed forward?

**Comment by k-takata** [comment]
It seems that my patch has been (partly) included in the prerelease version: http://www.drchip.org/astronaut/vim/#NETRW.
If @cecamp thinks it is ready for releasing, he will send the files to Bram. But he doesn't think it is ready?

**Comment by jerik** [comment]
Improved version of @strboul [suggestion](https://github.com/vim/vim/issues/4738#issuecomment-714609892). 
* Open the url under the cursor, not the first url in the line
* Fixes issues with ? in url, that causes to not open the url

```viml
if has('macunix')
  function! OpenURLUnderCursor()
    let s:uri = expand('<cWORD>')
    let s:uri = substitute(s:uri, '?', '\\?', '')
    let s:uri = shellescape(s:uri, 1)
    if s:uri != ''
      silent exec "!open '".s:uri."'"
      :redraw!
    endif
  endfunction
  nnoremap gx :call OpenURLUnderCursor()<CR>
endif
```

**Comment by make-github-pseudonymous-again** [comment]
> Improved version of @strboul [suggestion](https://github.com/vim/vim/issues/4738#issuecomment-714609892).
> 
>     * Open the url under the cursor, not the first url in the line
> 
>     * Fixes issues with ? in url, that causes to not open the url
> 
> 
> ```viml
> if has('macunix')
>   function! OpenURLUnderCursor()
>     let s:uri = expand('<cWORD>')
>     let s:uri = substitute(s:uri, '?', '\\?', '')
>     let s:uri = shellescape(s:uri, 1)
>     if s:uri != ''
>       silent exec "!open '".s:uri."'"
>       :redraw!
>     endif
>   endfunction
>   nnoremap gx :call OpenURLUnderCursor()<CR>
> endif
> ```

@jerik Does not work when the URL is in parentheses.

Also, both in the original function and your version the use of `shellescape` and single quotes around `s:uri` is redundant. With parentheses, using the `fish` shell, this causes the url to be executed as a command.

Fortunately you can combine both solutions to get something that sort-of works:
```viml
if has('macunix')
  function! OpenURLUnderCursor()
    let s:uri = expand('<cWORD>')
    let s:uri = matchstr(s:uri, '[a-z]*:\/\/[^ >,;()]*')
    let s:uri = substitute(s:uri, '?', '\\?', '')
    let s:uri = shellescape(s:uri, 1)
    if s:uri != ''
      silent exec "!open '".s:uri."'"
      :redraw!
    endif
  endfunction
  nnoremap gx :call OpenURLUnderCursor()<CR>
endif
```

But this does not solve the problem in general since parentheses are valid in urls.

**Comment by imambungo** [comment]
- Improved regex to match URL inside single or double quotes
- Fixes issues with # in URL, that causes to not open the URL

```vim
function! OpenURLUnderCursor()
	let s:uri = expand('<cWORD>')
	let s:uri = matchstr(s:uri, "[a-z]*:\/\/[^ >,;)'\"]*")
	let s:uri = substitute(s:uri, '#', '\\#', '')
	" let s:uri = substitute(s:uri, '?', '\\?', '')
	" let s:uri = shellescape(s:uri, 1)
	if s:uri != ''
		silent exec "!xdg-open '".s:uri."' > /dev/null"
		:redraw!
	endif
endfunction
nnoremap gx :call OpenURLUnderCursor()<CR>
```
I don't need the commented lines in zsh.

About the `> /dev/null` redirection, [this](https://user-images.githubusercontent.com/34421249/111047048-7e52ca00-847f-11eb-8780-424cc5855ce8.png) is what happens on my machine (ubuntu) without it:

![image](https://user-images.githubusercontent.com/34421249/111047048-7e52ca00-847f-11eb-8780-424cc5855ce8.png)

That happens when the default browser is chrome.

**Comment by mcepl** [comment]
And now it works again (in neovim). Damn!

**Comment by SeniorMars** [comment]
I just tested this again today, and it seems I still get this error on linux. The suggestions in this chat work, but I wish it would be patchd.

**Comment by eirnym** [comment]
Neovim has its own fork?

**Comment by mcepl** [comment]
> Neovim has its own fork?

No, it just periodically refreshes by applying the particular vim patch (as usual, neovim tries hard not to fork).

**Comment by ghost** [comment]
really, it is not yet fixed ? I am astonished to continue to receive mail about this. Honestly, it makes me sad to see how broken part of viM is.

such feature should have its own C implementation in the main codebase than viM. We have to stop to rely on brittle vimscript.

I have to compile my own vim (ok it's fast) and use the right vimscript version but come on, it is not sustainable.

all the best, pal!

**Comment by chrisbra** [comment]
> such feature should have its own C implementation in the main codebase than viM. We have to stop to rely on brittle vimscript.

In the spirit of open source, you may contribute and become famous :)

**Comment by ghost** [comment]
> > such feature should have its own C implementation in the main codebase than viM. We have to stop to rely on brittle vimscript.
> 
> In the spirit of open source, you may contribute and become famous :)

funny, there is  especially this quote in netrwPlugin.vim file
_"  But be doers of the Word, and not only hearers, deluding your own selves"
 (James 1:22 RSV)_

I am a vim user, not vim maintainer, I can't afford to learn this codebase right now, netwr is a 13k LoC beast. :/

do you know if there were discussion about a proper file manger implementation for vim?

**Comment by Shane-XB-Qian** [comment]
https://github.com/vim/vim/issues/4738#issuecomment-565330300

neednot 13k LoC, but just remap gx to whatever you like, sometime the style did not match everyone like, specially looks vi vim born with Linux..
Since many Lang api or just some shell cmd offered way to open link, as long as that link can be opened.
you can :
1, let g:netrw_nogx = 1
2, re-map gx to whatever you like.
3, done.
// and someone above offered patch as well.

// and this ticket looks kept too long... :)

**Comment by ghost** [comment]
you missed the point :) 
I do have a workaround, that's not the question. Here, we see that netrw is just not reliable.

i agree on the last top sentence.

**Comment by mcepl** [comment]
> Since many Lang api or just some shell cmd offered way to open link, as long as that link can be opened.
> you can :
> 1, let g:netrw_nogx = 1
> 2, re-map gx to whatever you like.

Thank you for this:
1. I missed `g:netrw_nogx`, which is useful
2. I don’t think `<cword>` is the right expression. `<cfile>` seems to be more appropriate.

So, my final solution is:
```vim
" Doesn't work because of gh#vim/vim#4738
" let g:netrw_browsex_viewer='setsid osurl'
let g:netrw_nogx=1

" This is just temporary workaround until the above issue is truly
" resolved.
" https://github.com/vim/vim/issues/4738#issuecomment-830820565
" https://bugzilla.suse.com/show_bug.cgi?id=1173583
" (https://bugzilla.suse.com/show_bug.cgi?id=1173583)
" gh#vim/vim#4738
function! OpenURLUnderCursor()
  let s:uri = expand('<cfile>')
  echom "s:uri = " . s:uri
  if s:uri != ''
    silent exec "!osurl '".s:uri."'"
    :redraw!
  endif
endfunction
nnoremap gx :call OpenURLUnderCursor()<CR>
```
(of course, you should replace calling [`osurl`](https://gitlab.com/mcepl/vim-suse-changes/-/blob/master/osurl) in my case with anything you want).

**Comment by k-takata** [comment]
I recommend [tyru/open-browser.vim](https://github.com/tyru/open-browser.vim).

**Comment by vhristev** [comment]
OS: MacOSX 11.2.3
NeoVim: NVIM v0.5.0-dev+1318-g61aefaf29

That was working for me for the past years but now it seems after some upgrade and I upgrade the entire system + neovim + all plugins. I end up here. 

When i use 'GBrowse' i get an error below:
`  **warning** (netrw) shell signalled an error
`

Strange is that URL looks fine if i copy it and paste it in the browser is working. 

'gx' also does not work. I tried setting 

let g:netrw_http_cmd='open' but without success.

**Comment by mcepl** [comment]
> let g:netrw_http_cmd='open' but without success.

Not sure about `:GBrowse`, but otherwise I just gave up ... see https://github.com/vim/vim/issues/4738#issuecomment-830831665 .

**Comment by lifepillar** [comment]
I haven't been using netrw for a long time (at least with macOS), but `gx` was useful, so I retained it.  In my `vimrc` I have something like the following, which works for me in Vim 8.2.2874 and macOS 11.3.1:

```vim
filetype plugin on
let g:loaded_netrwPlugin = 1
nnoremap gx :call netrw#BrowseX(expand((exists("g:netrw_gx")? g:netrw_gx : '<cfile>')),netrw#CheckIfRemote())<cr>
```

Try it with `vim -N -i NONE -u gx.vim`, where `gx.vim` is the script above. At the time, I had copied the definition for `gx` from netrw's source code. Comparing with the current version, the only difference I see is that `netrw#GX()` has been added.

Removing `netrw#GX()` from the argument of `netrw#CheckIfRemote()` in `$VIMRUNTIME/plugin/netrwPlugin.vim`, line 84, seems to fix the issue (Vim 8.2.2874). No idea what the consequences of this change are, though.

**Comment by y** [comment]
Is it expected behavior that using 'gx' will cause a modified file to be saved when 'autowrite' is set?

**Comment by rmagillxyz** [comment]
> I haven't been using netrw for a long time (at least with macOS), but `gx` was useful, so I retained it. In my `vimrc` I have something like the following, which works for me in Vim 8.2.2874 and macOS 11.3.1:
> 
> ```viml
> plugin on
> let g:loaded_netrwPlugin = 1
> nnoremap gx :call netrw#BrowseX(expand((exists("g:netrw_gx")? g:netrw_gx : '<cfile>')),netrw#CheckIfRemote())<cr>
> ```
> 
> Try it with `vim -N -i NONE -u gx.vim`, where `gx.vim` is the script above. At the time, I had copied the definition for `gx` from netrw's source code. Comparing with the current version, the only difference I see is that `netrw#GX()` has been added.
> 
> Removing `netrw#GX()` from the argument of `netrw#CheckIfRemote()` in `$VIMRUNTIME/plugin/netrwPlugin.vim`, line 84, seems to fix the issue (Vim 8.2.2874). No idea what the consequences of this change are, though.

@lifepillar  I've been messing with this for a while, and found this thread from a thread pointing here which was created in 2017 and starting to think I would never figure it out but alas! thanks mate. 

Also,   here is the code in lua for anyone switching over: 
```lua
vim.cmd('filetype plugin on')
vim.g.loaded_netrwPlugin = 1
vim.api.nvim_set_keymap('n', 'gx', ":call netrw#BrowseX(expand((exists('g:netrw_gx')? g:netrw_gx : '<cfile>')),netrw#CheckIfRemote())<cr>", {noremap = true, silent = true})
```

**Comment by felipec** [comment]
> really, it is not yet fixed ? I am astonished to continue to receive mail about this. Honestly, it makes me sad to see how broken part of viM is.

That's because vim has an archaic development process, and in particular parts of code maintained by Charles E. Campbell ([web site](http://www.drchip.org/)), who is not precisely responsive.

It took me **six years** to get a patch merged to vim because he is the maintainer of `sh.vim`. ([I sent the first patch in 2014](https://groups.google.com/g/vim_dev/c/DH6gRqn9KBk/m/gPmD4zRKPMUJ)).

I could probably dig deep into the code and find the proper fix, but why bother if the patch is going to be ignored? (just like my [last patch](https://groups.google.com/g/vim_dev/c/CS5LccGEQQY/m/4Q9VGUELAgAJ))

It's easier to just give on that software and find a workaround.

**Comment by rmagillxyz** [comment]
> > really, it is not yet fixed ? I am astonished to continue to receive mail about this. Honestly, it makes me sad to see how broken part of viM is.
> 
> That's because vim has an archaic development process, and in particular parts of code maintained by Charles E. Campbell ([web site](http://www.drchip.org/)), who is not precisely responsive.
> 
> It took me **six years** to get a patch merged to vim because he is the maintainer of `sh.vim`. ([I sent the first patch in 2014](https://groups.google.com/g/vim_dev/c/DH6gRqn9KBk/m/gPmD4zRKPMUJ)).
> 
> I could probably dig deep into the code and find the proper fix, but why bother if the patch is going to be ignored? (just like my [last patch](https://groups.google.com/g/vim_dev/c/CS5LccGEQQY/m/4Q9VGUELAgAJ))
> 
> It's easier to just give on that software and find a workaround.

What's your opinion of vim vs neovim? Neovim is much more active.

**Comment by mcepl** [comment]
> What's your opinion of vim vs neovim? Neovim is much more active.

I use neovim, but it doesn’t help you here (or with `sh.vim`), because neovim uses runtime (i.e., non-C files) from vim, and tries very hard not to make much difference between the two unless necessary. So, for example, this is the neovim ticket as well.

**Comment by felipec** [comment]
> What's your opinion of vim vs neovim? Neovim is much more active.

I love the idea of neovim, but I need  GUI, and I haven't found any suitable one, so I'm forced to use gvim (until I port it to use nvim).

Not that this matters on this particular case, since the problem is the runtime.

**Comment by rmagillxyz** [comment]
> > What's your opinion of vim vs neovim? Neovim is much more active.
> 
> I use neovim, but it doesn’t help you here (or with `sh.vim`), because neovim uses runtime (i.e., non-C files) from vim, and tries very hard not to make much difference between the two unless necessary. So, for example, this is the neovim ticket as well.

maybe once it becomes more stable, they will be willing to dive back into the more archaic code.

**Comment by rmagillxyz** [comment]
> > What's your opinion of vim vs neovim? Neovim is much more active.
> 
> I love the idea of neovim, but I need GUI, and I haven't found any suitable one, so I'm forced to use gvim (until I port it to use nvim).
> 
> Not that this matters on this particular case, since the problem is the runtime.

you could also use vscodium running neovim on the backend. I've never used the config, but this project has the vscodium config built into it. Check out the README.

**Comment by sroracle** [comment]
Please keep the discussion on-topic to the bug, you are spamming many people's inboxes.

**Comment by felipec** [comment]
Does spending several hours investigating the history of the code and fixing the issue afford me one comment?

If you don't want people to "spam inboxes", then provide a place where people can freely discuss these issues, like a Gitter chatroom (see request #8291).

---

Anyway, I sent the patch with the fix to the mailing list: [netrw: make gx work again](https://groups.google.com/g/vim_dev/c/UxR-k2D-f1Y/m/5gGiX-p6BwAJ).

The proper way to workaround the bug is:

```vim
nnoremap gx :call netrw#BrowseX(netrw#GX(),0)<cr>
```

The regression was caused by commit 85850f3a5 v8.1.1714.

Note that this commit breaks all the rules of proper coding practices.

1. It's not a small logically separate change ([Commit Guidelines](https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project)), it's a **HUGE** dump of 7,570 lines of code. Not only does it mix code for many plugins, but it updates **three years** of netrw development into a single patch.
2. It doesn't have attribution. We don't know who did what (nor when for that matter).
3. It doesn't explain **why** the changes were made.

`git bisect` makes it pretty easy to find the commit that causes a regression, but if the commit is 8,000 lines long, that doesn't help much.

It took me a while to figure out what change was responsible for the breakage: `CheckIfRemote()` was changed to `CheckIfRemote(netrw#GX())`. Why? We will never know, because the development of netrw is completely opaque.

Once I fixed the issue I checked to see if Charles Campbell's code fixed it in a different way. First I had to wait a long time because his server wasn't responding. Once I downloaded the file it turns out it's a `.vba.gz` which requires yet another plugin of Campbell to extract (it's a vimball). His plugin extracted the files into my `$HOME` directory, which I didn't want to be messed up. But then finally I was able to see the diff.

Such a joy.

Anyway, I isolated and back-ported his fix into the master branch of vim.

Now let's see how many years they take to do something with it.

[0001-netrw-make-gx-work-again.txt](https://github.com/vim/vim/files/6566881/0001-netrw-make-gx-work-again.txt)

**Comment by 9mm** [comment]
@felipec amazing

**Comment by nickspoons** [comment]
@filipec you seem to be working quite hard to not understand how the runtime file maintenance works, so here's a summary.

Runtime files and vim's few included plugins are mostly not maintained by @brammool or the vim community in general, but by individual maintainers like @cecamp. This is, generally speaking, because the runtime files fall outside the scope of the main project, many of them require domain-specific knowledge, and to divide responsibilities.

When maintainers are satisfied with changes to a runtime file, then send it to Bram for inclusion in the main repo. This happens constantly, and Bram gathers up the changes and commits them here at various intervals. These runtime file commits are **not** granular, they are not supposed to be. For granular history, go to the maintainer.

You are clearly frustrated, and perhaps with good reason, but your aggressive and insulting writing style does not make people want to spend their time on the issues you raise.

**Comment by felipec** [comment]
> @filipec you seem to be working quite hard to not understand how the runtime file maintenance works, so here's a summary.

I know perfectly well how the runtime maintenance works, but more importantly: I know how it **should** work, and it's not like this.

> Runtime files and vim's few included plugins are mostly not maintained by @brammool or the vim community in general, but by individual maintainers like @cecamp. This is, generally speaking, because the runtime files fall outside the scope of the main project, many of them require domain-specific knowledge, and to divide responsibilities.

And that's why they should be maintained a in a separate repository.

> These runtime file commits are **not** granular, they are not supposed to be.

But they **should be**. It's the recommended coding practice everywhere. Do I really need to list the reasons why?

> For granular history, go to the maintainer.

Bram **is** the maintainer. He is the person responsible for the vim repository. He needs to demand granular history from the submaintainers, like plenty of other maintainers in other repositories.

If he doesn't want to be responsible, then he needs to split the vim-runtime, and assign another person responsible for that project. But until he does that, he is in charge.

In any company the manager is responsible for all his/her subordinates. You can't shift the blame to somebody under you.

Responsibility is a hierarchy.

If a new version of Linux corrupts your file system do you think Linus Torvalds would say "don't look at me, that's the responsibility of that *other* guy". No. When he tags Linux 5.12, he assumes responsibility of **everything** that is part of that release.

With great power comes great responsibility.

> You are clearly frustrated, and perhaps with good reason, but your aggressive and insulting writing style does not make people want to spend their time on the issues you raise.

I did not raise the issue, the issue was raised by about 30 people **two years ago**. I only became aware of it 3 days ago.

And don't slander me. I never insulted anyone. Show me where I did that.

P.S. It's @felipec, not @filipec

**Comment by felipec** [comment]
Apparently the fix I backported was already sent as a pull request in pull #7188. Charles Campbell already picked it up for his repository, but has not sent it to @brammool.

**Comment by k-takata** [comment]
I have already written that before in https://github.com/vim/vim/issues/4738#issuecomment-755926374.
Someone who is bothered by this issue should try the [prerelease version](http://www.drchip.org/astronaut/vim/#NETRW) first.
Another way I recommend is using [tyru/open-browser.vim](https://github.com/tyru/open-browser.vim).

**Comment by Shane-XB-Qian** [comment]
```
	v168:	Dec 12, 2019	* scp timeout error message not reported,
				  hopefully now fixed (Shane Xb Qian)
```
@felipec i raised patch to netrw too in the past, it was not such hard actually... i got response from cecamp shortly.
if you want to check this issue deeply, perhaps can try to follow the way as he said in this ticket beginning................
note: he is not only the author of this netrw, but also lots runtime (of vim), you can image his workload......................

there is a IRC room (compared to gitter) if you want IM chat, you can find it in the vim.org.

i did not check the detail of why this change happened, but mostly i can image is someone like getting 'http://foo.org/bar.txt' into vim buffer directly (vs opened into a browser).

Wish now it is clear enough............................................

**Comment by felipec** [comment]
> I have already written that before in [#4738 (comment)](https://github.com/vim/vim/issues/4738#issuecomment-755926374).

Ahh, apparently GitHub was hiding 80 comments that I didn't see originally. The discussion is bigger than I thought.

I found the issue with `xdg-open`, and it happens with anything that uses `posix_spawn()`, so it's definitely a vim issue. My guess is that the problem is with `TIOCSCTTY`:

```c
# ifdef TIOCSCTTY
 // Try to become controlling tty (probably doesn't work,
 // unless run by root)
 ioctl(pty_slave_fd, TIOCSCTTY, (char *)NULL);
# endif
```

Removing that code makes everything work. I already [sent a patch to the ML](https://groups.google.com/g/vim_dev/c/gPu_N2LhmK4).

One way you can disable this stuff is by doing `set noshelltemp`. With that `xdg-open` works fine.

Who knows why that code was added: it's there since the first git commit in 2004, and hasn't been changed since.

**Comment by k-takata** [comment]
(Off topic)
> Who knows why that code was added: it's there since the first git commit in 2004, and hasn't been changed since.

You may want to try https://github.com/vim/vim-history.

**Comment by 9mm** [comment]
Damn... 21 year old commits. I was just a wee lad sipping cool aid worried about Y2K while grownups drank champagne. And at that very moment, vim commits were made that would forever shape my future, 10 years before I even discovered vim 😳

**Comment by felipec** [comment]
> You may want to try https://github.com/vim/vim-history.

Very nice! Unfortunately that still doesn't help. All we have is a huge commit from v3.0 to v4.0 in 1996 that added TIOCSCTTY for some reason.

**Comment by brammool** [comment]
> > You may want to try https://github.com/vim/vim-history.
> 
> Very nice! Unfortunately that still doesn't help. All we have is a
> huge commit from v3.0 to v4.0 in 1996 that added TIOCSCTTY for some
> reason.

I also don't recall why it was added (and not going to spend time
searching old mail), and the remark says it might not even work.
Since it's been shown that removing it fixes a problem, and there is no
clue what problem removing it would cause, we can drop those lines.
In case someone runs into a problem we'll find out.

Would be nice if we can test this somehow.


-- 
User:       I'm having problems with my text editor.
Help desk:  Which editor are you using?
User:       I don't know, but it's version VI (pronounced: 6).
Help desk:  Oh, then you should upgrade to version VIM (pronounced: 994).

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by felipec** [comment]
@brammool I did send a test to the mailing list: [test program](https://groups.google.com/g/vim_dev/c/gPu_N2LhmK4/m/JhokOrmfBwAJ).

```c
#include <spawn.h>
#include <signal.h>

extern char **environ;

int main(void)
{
	char *const argv[] = { "/usr/bin/chromium", "http://google.com", NULL };
	posix_spawnattr_t attr;
	sigset_t mask;

	posix_spawnattr_init(&attr);
	sigemptyset(&mask);
	sigaddset(&mask, SIGHUP);

	posix_spawnattr_setsigdefault(&attr, &mask);
	posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGDEF);

	return posix_spawn(NULL, argv[0], NULL, &attr, argv, environ);
}
```

```sh
gcc test.c -o test
vim -g -c ':!./test'
```

Doesn't do anything unless you apply my patch or add `set noshelltemp` to your vimrc.

**Comment by brammool** [comment]
> @brammool I did send a test to the mailing list: [test program](https://groups.google.com/g/vim_dev/c/gPu_N2LhmK4/m/JhokOrmfBwAJ).
> 
> ```c
> #include <spawn.h>
> #include <signal.h>
> 
> extern char **environ;
> 
> int main(void)
> {
> 	char *const argv[] = { "/usr/bin/chromium", "http://google.com", NULL };
> 	posix_spawnattr_t attr;
> 	sigset_t mask;
> 
> 	posix_spawnattr_init(&attr);
> 	sigemptyset(&mask);
> 	sigaddset(&mask, SIGHUP);
> 
> 	posix_spawnattr_setsigdefault(&attr, &mask);
> 	posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGDEF);
> 
> 	return posix_spawn(NULL, argv[0], NULL, &attr, argv, environ);
> }
> ```
> 
> ```sh
> gcc test.c -o test
> vim -g -c ':!./test'
> ```
> 
> Doesn't do anything unless you apply my patch or add `set noshelltemp`
> to your vimrc.

That is a manual test.  It doesn't work as-is on my system (I don't have
chromium) and requires looking at what happens.  Not really fit for an
automated test.

I suppose it does require using an external command.  Compiling C code
can fail in many ways (not sure where spawn.h is available).  Isn't
there an existing command we can use?  Ideally it would create some
output we can check for.

-- 
BEDEVERE: Look!  It's the old man from scene 24 - what's he Doing here?
ARTHUR:   He is the keeper of the Bridge.  He asks each traveler five
          questions ...
GALAHAD:  Three questions.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by felipec** [comment]
> That is a manual test.

That's because nobody said they wanted an automated test.

> It doesn't work as-is on my system (I don't have chromium)

chromium was an **example**. You can try literally anything.

> and requires looking at what happens.

No, here:

```bash
#!/bin/bash

cat > test.c <<EOF &&
#include <spawn.h>
#include <signal.h>

extern char **environ;

int main(int argc, char *argv[])
{
	posix_spawnattr_t attr;
	sigset_t mask;

	posix_spawnattr_init(&attr);
	sigemptyset(&mask);
	sigaddset(&mask, SIGHUP);

	posix_spawnattr_setsigdefault(&attr, &mask);
	posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGDEF);

	return posix_spawn(NULL, argv[1], NULL, &attr, argv + 1, environ);
}
EOF
gcc test.c -o test &&
vim -g --nofork -c $':!./test /usr/bin/touch /tmp/test-output\n:q' &&
! test -f /tmp/test-output && echo "FAILED!"
```

> Not really fit for an automated test. I suppose it does require using an external command. Compiling C code can fail in many ways (not sure where spawn.h is available).

It should be trivial to look at any posix_spawn implementation and remove anything that isn't relevant for the test.

But at this point I'm doing everything, when it literally takes two seconds to reproduce the issue with the code above.

> Isn't there an existing command we can use?

Perhaps all we need is a command that resets the SIGHUP signal to the default handler, but at least I couldn't reproduce that with bash.

>  Ideally it would create some output we can check for.

The command above does create an output.

**Comment by felipec** [comment]
@brammool Aha! Found a simple way to reproduce it:

```sh
vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'
```
  
No file is created.

**Comment by brammool** [comment]
> @brammool Aha! Found a simple way to reproduce it:
> 
> ```sh
> vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'
> ```
>   
> No file is created.

Unfortunately for me this does create the file.  I tried a few variants,
adding "--clean" and explicitly setting the shell, but it still created
the file.

-- 
BRIDGEKEEPER: What is the air-speed velocity of an unladen swallow?
ARTHUR:       What do you mean?  An African or European swallow?
BRIDGEKEEPER: Er ...  I don't know that ... Aaaaarrrrrrggghhh!
   BRIDGEKEEPER is cast into the gorge.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by eirnym** [comment]
Probably it depends on how it build. In some systems it uses posix spawn, on some not. 

I'd suggest to put more details on system used by @felipec to reproduce an issue and properly create a regression test for it. "Not every Linux distro is the same" as not every cat is black at night

**Comment by felipec** [comment]
@eirnym I'm using Arch Linux, the way they build is nothing special: [PKGBUILD](https://github.com/archlinux/svntogit-packages/blob/packages/vim/trunk/PKGBUILD). But I'm also building vim myself, and all I configured is `--enable-gui=gtk3`.

I don't think it's due to the build: I found another moving part, it only happens with zsh, not with bash.

This works:

    SHELL=bash vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'
    
This doesn't:

    SHELL=zsh vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'

**Comment by eirnym** [comment]
Thank you for information. I'm also use ArchLinux and zsh. My guess, the touch in zsh is a built-in. 

Could you try to specify full path?

**Comment by BachoSeven** [comment]
It is not a built-in though:
``` sh
  echo $SHELL && which touch 
/usr/bin/zsh 
/usr/bin/touch 
```

**Comment by felipec** [comment]
> Could you try to specify full path?

Yes I have done that, it doesn't work.

It's not a matter of builtins as it happens with the compiled program I provided that uses posix_spawn().

The issue is about spawning background processes. I opened a new issue #8314 since this is not related to `gx` (that is yet another issue).

**Comment by brammool** [comment]
> @eirnym I'm using Arch Linux, the way they build is nothing special: [PKGBUILD](https://github.com/archlinux/svntogit-packages/blob/packages/vim/trunk/PKGBUILD). But I'm also building vim myself, and all I configured is `--enable-gui=gtk3`.
> 
> I don't think it's due to the build: I found another moving part, it only happens with zsh, not with bash.
> 
> This works:
> 
>     SHELL=bash vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'
>     
> This doesn't:
> 
>     SHELL=zsh vim -g -c 'set shelltemp' -c ':!touch /tmp/foobar &'

OK, so it depends on how the shell starts the command.  Unfortunately
that makes it less useful for a test.  I don't even have zsh installed.

As mentioned before, including a C program to reproduce the failure is a
hassle.  It would probably require some configure changes to make sure
it works everywhere.  I tend to think we can omit adding a test, since
it would be too specific to be useful.

-- 
Kisses may last for as much as, but no more than, five minutes.
		[real standing law in Iowa, United States of America]

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by felipec** [comment]
> OK, so it depends on how the shell starts the command. Unfortunately that makes it less useful for a test.

Before trying to write an **automated** test we first need to understand what the actual issue is. And the first step in doing that is for others to be able to reproduce the issue.

At the moment anyone interested in figuring that out needs to use `zsh`, and install it if not already installed. It takes one command (e.g. `pacman -S zsh`).

A way to reproduce without `zsh` could be found afterwards.

**Comment by felipec** [comment]
I found yet another issue with netrw not opening URLs with `@` and I decided enough is enough. I started removing all the code that is not useful, and in the end, I arrived at only **one** line of code:

```vim
nnoremap <silent> gx :execute 'silent! !xdg-open ' . shellescape(expand('<cWORD>'), 1)<cr>
```

That not only does everything that netrw does, but it does it better.

To support `g:netrw_browsex_viewer` for backwards compatibility and some default openers such as `xdg-open` and `open` a few more lines are needed, so I decided to create a new simple plugin: [vim-sanegx](https://github.com/felipec/vim-sanegx).

```sh
git clone https://github.com/felipec/vim-sanegx.git ~/.vim/pack/plugins/start/sanegx
```

There's no need to wait for others to do the sane thing. If you have had issues with `xdg-open` in the past, so `set noshelltemp` to avoid a bug in gvim.

**Comment by mcepl** [comment]
@brammool And on that note, could somebody move https://github.com/vim/vim/pull/7188 to some conclusion, please? I have the patch in my build of neovim since that October 2020 and it works just fine.

**Comment by k-takata** [comment]
The status is not changed from [my old comment](https://github.com/vim/vim/issues/4738#issuecomment-755926374).

It seems that my patch has been (partly) included in the prerelease version: http://www.drchip.org/astronaut/vim/#NETRW.
So, someone who is bothered by this issue should try the prerelease version first.
Once Charles thinks it is ready for release, he will send the files to Bram.
ping @cecamp 

Another way I recommend is using tyru/open-browser.vim.

**Comment by GianniGi** [comment]
I have no idea on how to solve this on windows 10...

**Comment by chrisbra** [comment]
please provide more information, version of vim and netrw shell used, gui, etc and reproduction steps

**Comment by felipec** [comment]
@GianniGi Use [vim-sanegx](https://github.com/felipec/vim-sanegx)?

**Comment by mcepl** [comment]
> I have no idea on how to solve this on windows 10...

If you have reasonably recent vim (or neovim), then this issue should be actually fixed, so you should probably write a new ticket and fill in all information about version of your vim and exactly what you tried to achieve and what actually happened.

**Comment by GianniGi** [comment]
> please provide more information, version of vim and netrw shell used, gui, etc and reproduction steps

Thanks, I solved by installing the latest vim version and by running this:

:let g:netrw_browsex_viewer= "open"

VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 25 2022 23:04:24)
MS-Windows 64-bit console version
Included patches: 1-4219
Compiled by appveyor@APPVYR-WIN

@felipec 
@mcepl

**Comment by GianniGi** [comment]
Reinstalled vim to the latest version 9.0 updated all plugins. Nothing works anymore. (Well, it didn't work even before).

gx is broken

VIM - Vi IMproved 9.0 (2022 Jun 28, compilato Sep 18 2022 22:03:46)
Versione MS-Windows 64-bit console
Patch incluse: 1-500

**Comment by igbanam** [comment]
`gx` seems to work perfectly fine for me in both use cases, @GianniGi 

**Case 1**
- `:Explore`
- Select any file
- `x`
- _this opens the right app_

**Case 2**
- Open an empty buffer
- Insert "https://google.com"
- With the cursor within the URL and `gx`
- _this opens the browser_

When I just installed Vim 9, looked like the editor had a meltdown as well. But when I fixed the vim9script + discrepancies with my terminal, everything works just fine …even better than before

**Comment by igbanam** [comment]
For context, I'm on a Mac M1

```
$ vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Sep 01 2022 15:00:53)
macOS version - arm64
Included patches: 1-350
```

**Comment by k-takata** [comment]
@GianniGi How did you test that?

I tested the following steps with Vim 9.0.0525 on Windows and worked fine:

1. Run vim with a clean setting: `vim --clean --cmd "set loadplugins"`
2. Insert `https://google.com`.
3. Type `gx`.

**Comment by GianniGi** [comment]
> 1. vim --clean --cmd "set loadplugins"

.... I had no idea that vim --clean was loading the plugins anyway

**Comment by chrisbra** [comment]
>>        vim --clean --cmd "set loadplugins"

> .... I had no idea that vim --clean was loading the plugins anyway

`vim --clean` does not load any plugins, that's why you need to enable loading of global plugins using `:set loadplugins`.


> Now I have to find the offending plugin??? maybe something markdown related???

Check your mappings, `:verbose map gx`

**Comment by GianniGi** [comment]
> > Now I have to find the offending plugin??? maybe something markdown related???

Hello, the problem was the coc plugin, somehow it messed up even markdown files.

**Comment by iamaamir** [comment]
i am on macOS using vim 9.0.900 and facing the same
<img width="566" alt="image" src="https://user-images.githubusercontent.com/8420386/203841446-c83b2a30-4a7b-48fd-b871-9b4c69ffba7d.png">

**Comment by cecamp** [comment]
Does the gx problem still occur with netrw v172g?  (http://www.drchip.org/astronaut/vim/index.html#NETRW)
If it does, please follow the direction in :help netrw-debug and send me a trace.

Thank you,
Chip Campbell
